### ISSUE TITLE
Custom message attributes

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#345]
In some situation, it can be useful to have the ability to use custom attributes in WAMP message details/options.

The spec defines a set of attributes (depending on message type) for details/options we agreed on .. kinda "official" ones. However, we could allow custom/implementation defined attributes in principle (in details/options).

I think it was discussed before (somewhere I forgot .. maybe on an issue here, or the WAMP mailing list) and was raised recently again as a PR on ABJS https://github.com/crossbario/autobahn-js/pull/401

I also think from that discussion we concluded that having a prefix `x_` for custom/implementation defined attributes would be could:

* to prevent collisions with official attributes
* clearly mark custom attributes
* potentially disallow (eg in router) those using 1 kill switch

FWIW, Autobahn-Python makes use of that (not yet agreed and under-defined) approach using the following regex:

```
_CUSTOM_ATTRIBUTE = re.compile(r"^x_([a-z][0-9a-z_]+)?$")
```

pls see: https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/message.py#L90

this needs consensus among implementors. I think it would be better to agree on an approach, rather then letting this loose end further proliferate into different approaches.

---

I am +1 on:

* allow (configurable if the impl wants) custom attributes in WAMP message `details|options`
* use a `x_` prefix for custom attributes

**Comment by KSDaemon** [comment]
I'm also +1 for having custom attributes. In wampy.js i also has a request for custom hello details.
Having clearly defined possibility for custom options would be good. 
Also, prefix for separating wamp attributes and custom seems to be a good idea.

**Comment by meejah** [comment]
It might be useful to write down (existing) use-cases for these

**Comment by karenpeng** [comment]
A common use case is that router needs to store and maintain information of each client. If we can send custom attributes upon hello, as it's the initial message from client, we could make use of this information later on.

I agree that we should standardize custom attributes configuration. And I'm +1 with the idea of prefixing it with `x_`, it's clear and no overriding existing attributes since only `x_` attribute is accepted. I don't quite follow the kill switch reasoning though.

I have some questions:

---
- Will `custom_attributes` be a new configuration option, which will be merged into message's details? 
- Inside `custom_attributes`, is the prefix `x_` defined by users, or added in the library afterwards?
   - If it's defined by users, we need to communicate that all keys should prefix `x_`, otherwise it will be dropped.
   - If it's behind the scene, we should make it clear that no overriding will ever happen.
- What message will have custom attributes? 
   - RPC methods call/register/pubsub etc are exposed and take options as argument, it seems that hello is the only use case?
   - If so, should we name it as `hello_custom_attributes` and only pass it to hello message details?

**Comment by oberstet** [comment]
> A common use case is that router needs to store and maintain information of each client. If we can send custom attributes upon hello, as it's the initial message from client, we could make use of this information later on.

Why not simply do a procedure call with the respective info as the first action of the client?

`x_` attributes are not supposed to carry general app payload. That would be a misuse in my view. 

The extensibility should _not_ be used for _app_ specific stuff, but for WAMP _implementation_ specific extensions.

I agree with @meejah : to clarify the boundary between valid use cases for `x_` vs "just use regular RPC/PubSub app payloads" is what we need to agree first.

The technical details are secondary and mostly trivial (or will fall out naturally once we clarified above).

My view. Let's discuss;)

> I don't quite follow the kill switch reasoning though.

The kill switch would be a WAMP router configuration _option_ set by the admin of the router disallowing any use of `x_` attributes.

At least for Crossbar.io, we would add such an option, but other WAMP router implementors would be free to make it "always on" (non-configurable).

Reasoning:

- security and control over what apps do
- WAMP currently uses dynamic typing, but there is a proposal for static typing (of the WAMP message "frame", not necessarily the WAMP message app payload): https://github.com/wamp-proto/wamp-proto/tree/master/flatbuffers

rgd the latter: laying out all details of that proposal is beyond the scope of this issue ..

---

> custom_attributes

there won't be such dict. if this feature is making it into the spec, we'll use the existing `options/details` to carry `x_` attributes

> is the key prefix x_ defined by users,

no, the prefix `x_` would be part of the spec and fixed

> What message will have custom attributes? 

all messages with `options` or `details`

**Comment by karenpeng** [comment]
Gotcha, so what this issue proposes is for library authors to extend WAMP implementation, not for library users to pass in additional info.

Library such as wampy.js and nexus accepts hello custom attributes from users, I believe the motive was to save users an extra RPC request, @KSDaemon correct me if i'm wrong.

What was proposed in this issue is different. It's mostly for extensibility, as in the [RFC](https://wamp-proto.org/_static/gen/wamp_latest.html#extensibility):
> 6.1 Extensibility
Some WAMP messages contain Options|dict or Details|dict elements. This allows for future extensibility and implementations that only provide subsets of functionality by ignoring unimplemented attributes...

I think from just reading the RFC, it's not clear what use case it's for. Since it's `custom` attribute, it certainly has openness. I wonder what an actual implementational extension use case would be like? 

Anyway thanks for explaining, and I'm fond of the idea to define what this feature is for, as well as some details like prefix. This would help refine the RFC.

**Comment by oberstet** [comment]
Recently, we (Autobahn) have some users that want to have custom attributes supported (see https://github.com/crossbario/autobahn-python/issues/1214). So we really need some spec text to have a robust/sustainable basis first ..

**Comment by LaurenceGA** [comment]
Not really sure I am just missing something but I don't really understand the proposed change to the spec.

> allow (configurable if the impl wants) custom attributes in WAMP message details|options

Isn't this essentially already in the spec?
crossbario/autobahn-python#1214 points out that in section 6.1 of the protocol it states:
> Attributes unknown to an implementation MUST be ignored.

So you can pass whatever you want through the options/details dicts and everyone should ignore it unless it's an "official" one or they expect it.
What's the proposed change? That it's configurable? (MUST -> MAY?)

>use a x_ prefix for custom attributes

Is this just a suggestion? What should protocol implementors do with this information? How to implementations need to change?
Just not seeing the goal/use cases here :confused:

**Comment by oberstet** [comment]
> Attributes unknown to an implementation MUST be ignored.

ignoring does not mean forwarding. an implementation (both client and router) may simply drop those attributes and _not_ forward them on the wire - that is compliant with the spec text

the rationale for "ignoring" is strictly this:

* we (WAMP community/implementors) agree on a new _standard_ attribute (eg `foobar`, not prefixed with `x_`)
* older implementations not supporting that attribute must ignore it - but since they don't understand the new attribute, "ignoring" (without forwarding) is the right thing to do here

whereas `x_` prefixed attributes must be _forwarded_ (not dropped) by a router.

eg `CALL.options.x_foobar` should be blindly forwarded by a router to `INVOCATION.details.x_foobar`

does that make sense?

**Comment by mbonneau** [comment]
> whereas x_ prefixed attributes must be forwarded (not dropped) by a router.
>
> eg CALL.options.x_foobar should be blindly forwarded by a router to INVOCATION.details.x_foobar
>
>does that make sense?

I don't think it makes sense to pass options through as details. In some use cases an implementation specific `CALL` could use `x_` prefixed options to signal the router.

In Thruway, we have implemented RPC hooks. A client is able the hook an existing RPC and intercept calls. The hook can signal the router to pass the RPC through normally, could service the request itself, or can call the original "hooked" RPC and modify the response. This is controlled via `x_` options.

Passing `x_` through from `CALL.options` to `INVOCATION.details` does not seem appropriate as a router may be passing information that was only intended for a router - possibly leaking information.

> `x_` attributes are not supposed to carry general app payload. That would be a misuse in my view.
>
> The extensibility should not be used for app specific stuff, but for WAMP implementation specific extensions.

I agree with this. The user should just change the RPC to handle this information through the args/argskw.

If there really is a need to support "meta" info, I would rather see `CALL.options.meta` that gets passed through as `INVOCATION.details.meta`. (But again, I don't think this promotes healthy coding habits as it is just app payload at that point)

I am +1 on `x_`-prefix for options/details that are implementation specific for communication strictly between the router and the client.

**Comment by LaurenceGA** [comment]
I see, thank you.
I think I was misunderstanding what you meant by "forwarding".
WAMP _messages_ aren't really ever forwarded. I.e. the router creates a new `INVOCATION` message, it doesn't forward the `CALL` message.

I was basically just going to roughly say what @mbonneau has just said. So 👍there.
To specifically reiterate:
>Passing x_ through from CALL.options to INVOCATION.details does not seem appropriate as a router may be passing information that was only intended for a router - possibly leaking information.

For me, options/details are conceptually just properties of whatever message they appear in, that inform the recipient of whether to or how to do special handling. Passing these details off to the client or anywhere else seems weird. I suppose a strong use case would go a long way though.

**Comment by oberstet** [comment]
> I am +1 on x_-prefix for options/details that are implementation specific for communication strictly between the router and the client.

Agreed, makes more sense.

So:

* a WAMP client library SHOULD support custom options in the library API. those are sent to the router as `x_` attributes in `options` on the wire
* a WAMP router implementation or WAMP client library that receives a `x_` attribute on the wire it doesn't understand MUST ignore it (that is already part of the spec (section 6.1)
* a WAMP router implementation MAY use custom options, sent as `x_` attributes in `details` on the wire

Better?

**Comment by KSDaemon** [comment]
My thoughts are mostly the same as @mbonneau already wrote. 
@oberstet Seems to be good.

Well, if we all receive many requests for custom attributes in different WAMP API calls, maybe we can add more information in the spec about how to deal with this, that we propose to use options dictionary in concrete API calls, instead of expanding options. Yes, I understand, that it is not about API itself, but as WAMP specification is basic and the only document about how WAMP works, why not extend it with such howto description. 
O, even better, we can add such topic to https://wamp-proto.org/ site, not to spec.

**Comment by aramallo** [comment]
@oberstet I am reviewing this issue as we would like to work on it on AB Py and JS.

So just a tiny correction on the above

* a WAMP client library SHOULD support custom options and details in the library API. 
    * Custom options are sent to the router as `x_` attributes in `options` on the wire. 
    * Custom details are sent to the client as `x_` attributes in `details` on the wire.
* a WAMP router implementation or WAMP client library that receives a `x_` attribute on the wire it doesn't understand MUST ignore it (that is already part of the spec (section 6.1)
* a WAMP router implementation MAY use custom details, sent as `x_` attributes in details on the wire

**Comment by aramallo** [comment]
BTW, another approach that we've been using is for the client and router to allow the user to define the attributes each message should allow. In this way we can drop the `x_` prefix as the list of allowed attributes is pre-defined by the union of the set define by the spec and the set defined by the user.

This has the benefit that the custom attributes can be validated by the same mechanism the client and router use to validate WAMP attributes.

For example in Bondy we do the following:

```
Extensions = {
    "call" : [
          max_retries, 
          deadline
    ]
}
```

The example above tells the message validator to allow `max_retries` and `deadline` as additional attributes for `CALL.Options`.

I would rather prefer this approach to the `x_` attributes. 

We even allow the following in Bondy:

```
Extensions = {
    "register" : [
          {invoke: ["qll", "qlls"]}
    ]
}
```

This last example tells the message validator to extend the value range of `REGISTER.Options.invoke` to also allow two additional load balancing strategies (queue_least_loaded, and queue_least_loaded_sample).

**Comment by aramallo** [comment]
BTW this capability is fundamental to be able to experiment with WAMP features while using Autobahn clients.

**Comment by oberstet** [comment]
@aramallo rgd the small addition to the text (the bullet list) above: sure, that's good IMO, the client also needs to receive unknown x_ details and forward those to the application. I think best would be to discussa concrete PR for respective new spec text describing the `x_` stuff.

one thing I again came back to quickly is:

> I agree with @meejah : to clarify the boundary between valid use cases for x_ vs "just use regular RPC/PubSub app payloads" is what we need to agree first on targeted use cases.

this, plus a better description of how the pieces would work together, eg basically the only situation that seems to make sense is:

1. User App Service 1: **knows** about `x_foobar` <=>
2. WAMP library A: **doesn't know** about `x_foobar` <=>
3. WAMP wire protocol <=>
4. WAMP router X: **knows** about `x_foobar`
5. WAMP wire protocol <=>
6. WAMP library B: **doesn't know** about `x_foobar` <=>
7. User App Service 2: **knows** about `x_foobar`

if in 4. the router doesn't know about `x_foobar`, the best it could do is "ignore", which means `x_foobar` will have no effect on any (other) wamp client connected to that router

from that perspective, it might be fair to say `x_foobar` is a custom router X attribute (details or options) which is ignored and passed-through (!) in client libraries A/B which are unaware of `x_foobar`, but known to the app services 1/2

right?

> another approach that we've been using is for the client and router to allow the user to define the attributes each message should allow.

fwiw, I don't think this is a good idea in WAMP, I'd rather just send LISP S-Expressions then;) In my eyes, a network protocol is exactly about defining limitations on the set of allowed messages, where the "limitations" make sense: carve out meaning from the void. other than that, people should just use LISP (my personal view is, that didn't work .. excactly because people like "meaning", and that comes from clever meaningful limitations)

----

in summary, I think at this point, we really need a PR to nail down the details, and to reach a consensus decision upon merging the PR into the spec ...

**Comment by aramallo** [comment]
@oberstet 

> I think best would be to discuss concrete PR for respective new spec text describing the x_ stuff. One thing I again came back to quickly is:
    > I agree with @meejah : to clarify the boundary between valid use cases for x_ vs "just use regular RPC/PubSub app payloads" is what we need to agree first on targeted use cases.

Yes, one should not confuse custom attributes (control plane) with application payload (data plane). Custom attributes should be about using control-plane router-specific features that might never be part of the WAMP spec or to allow router-specific extensions to the protocol e.g. my example above where we have additional load balancing algorithms for the `Call.Options.invoke` attribute.

> if in 4. the router doesn't know about x_foobar, the best it could do is "ignore", which means x_foobar will have no effect on any (other) wamp client connected to that router

Yes, we are on the same page here.
> In my eyes, a network protocol is exactly about defining limitations on the set of allowed messages,

Yes, and that is what my last proposal does! Exactly because `x_` attribute proposal is not limiting the set of allowed messages (I guess you meant the content of the messages e.g. attributes).  On the contrary,  the client would allow  any `x_` attribute to be send to the router (step 3 in your example). In my proposal the client would not do that, since we would be telling it which attributes are allowed in each direction.

So just to be clear and using again your example:

1. User App Service 1  **knows** about `x_foobar` and tells the WAMP library A about it (declared via config) <=> 
1. User App Service 2  **knows** about `x_foobar` and tells the WAMP library B about it (declared via config) <=>
1. WAMP library A: **knows** `x_foobar` (because step 1 declared it), all other unknown attributes are discarded <=>
1. WAMP wire protocol, only `x_foobar` is present <=>
1. WAMP router X: **knows** about `x_foobar` (if it doesn't then the attribute is discarded)
1. WAMP wire protocol, only `x_foobar` is present <=>
1. WAMP library B: **knows** `x_foobar` (because step 2 declared it) all other unknown attributes will be discarded <=>
1. User App Service 2: **knows** about x_foobar

The router might use a similar mechanism (for example if router and client are sharing a common WAMP lib) or might hardcode the attributes it knows. In any case this is router specific and doesn't change the logic above.

----

So because of the above, I think having the `x_` prefix is unnecessary as all custom attributes are pre-declared.

**Comment by aramallo** [comment]
> allowing custom attribute value in a standard attribute is another issue. what you could do is use a custom attribute Call.Options.x_invoke with custom attribute value "qll" and "qlls"

Yes, that is a good option. And doesn't require a change on the client to support that.

> why would the client library need a list or a schema of custom attributes for doing ^? using a flag to enable the feature and any name x_* seems simpler

Because you would be sending all the `x_*` attributes to the router which is unnecessary. I would prefer for the client to drop those which the user has not declared.

How about, we combine both approaches. We allow all `x_*` attributes (a WAMP spec definition). And then if a particular client/lib implementor wants to further control/filter wkch attributes are transmitted over the wire it can (certainly the approach that we are taking).

**Comment by oberstet** [comment]
ok, so rgd the one Q above coming from our discussion, considering

1. only custom attributes (eg `Call.Options.x_invoke`) with standard (`"roundrobin"`) or custom value (eg `"qll"`), but no custom values in standard attributes
2. both custom attributes and custom values in standard attributes (eg `Call.Options.invoke == "qll"`)

we want to support 1), but _not_ 2), right?

-----

> How about, we combine both approaches. We allow all x_* attributes (a WAMP spec definition). And then if a particular client/lib implementor wants to further control/filter wkch attributes are transmitted over the wire it can (certainly the approach that we are taking).

yes, that sounds good!

it only applies to the sending side. an implementation MAY send `x_*` attributes, none, all, or any subset, eg selected via some client config

a receiving side MUST ignore (forward to app for client library or just ignore for router) any `x_*` - and it MAY know about some `x_*`, while ignoring others. and which ones of the `x_*` it _could_ act on it actually does act on can be further subselected by some router config.

does that match what you had in mind?

**Comment by oberstet** [comment]
very good!  at this point, I think the whole thread above already has enough substance and detail to nail it into spec text ..

---

> And then if a particular client/lib implementor wants to further control/filter wkch attributes are transmitted over the wire it can (certainly the approach that we are taking).

have been thinking a bit about this .. we could support it in eg AutobahnJS by exposing callbacks for filtering custom attributes, eg `tx_filter_custom_attribute` and `rx_filter_custom_attribute`  which a user can override to filter custom attributes set by user code before sending to wire, and to filter custom attributes received from the wire before forwarding to the app code ...

**Comment by KSDaemon** [comment]
Hi all! My 5 percent, if you don't mind:
If to choose between proposed 2 options: i personally prefer `x_*` way. This is simpler, more robust for applications that uses wamp under the hood.

What do you think about moving all custom attributes 1 level deeper? I mean in our dictionaries with attributes and options created a key `custom` and allow to place any client specific flags or properties there. In this case it is simpler for WAMP implementations to validate WAMP messages syntax, less errors and bugs (no possible overriding system keys).

**Comment by aramallo** [comment]
I like @KSDaemon option 2 too!

**Comment by KSDaemon** [comment]
Decided to take a run at this again as I think it is one of the easy-to-implement and at the same time yet powerful features, that gives future extensibility for developers without the need to extend the protocol itself.

Kinda wrapping it up once again to save time for those who don't want to read the whole thread again :)

From the technical side, we are mostly looking positively at next variants:

1. Custom attributes are at the top-level of `Options/Details` and must be prepended by `x_` prefix.
   1. End user must provide custom attributes already prefixed and client library only checks them for `x_` prefix
   2. End user provides unprefixed attribute and client library automatically adds `x_` prefix
   3. Both variants above
4. One level deep `Options.custom|dict` / `Details.custom|dict`, that if present, must have all string keys (arbitrary name) and arbitrary values.

For this q we have to agree/vote on which way we go. Specifically creating a plain list for easy choice :)

1. `x_*` + allow prefixed attributes from client side
2. `x_*` + allow unprefixed attributes from client side
3. `x_*` + allow both prefixed and unprefixed attributes from client side
4. `custom|dict`
5. Both: `custom|dict` and `x_*` + allow prefixed attributes from client side
6. Both: `custom|dict` and `x_*` + allow unprefixed attributes from client side
7. All variants

I'm +1 for `5` variant: Both: `custom|dict` and `x_*` + allow prefixed attributes from the client side. I don't like magic happening behind my eyes.

If we need to choose only one way — then I don't have an opinion about which one. Because from an implementation point in one case, it will be easier to go with `custom dict` and in another with `x_*`.

One question here that came into my mind: if we go `both` way may some wamp client/router implement only one variant? Should it be treated and announced as `supported`? Because we may have a situation when `lib A` has implemented only `x_*` and announced custom attributes support during the handshake, and `lib B` that has implemented only `custom|dict` and announced custom attributes support during the handshake — but in runtime, something went wrong :)

From the protocol descriptive side, we need to give really useful examples of how this feature may be used kinda best practice.

The one that I think is really good — was given by @ecorm in https://github.com/wamp-proto/wamp-proto/discussions/465 — implementing streams on top of progressive calls/invocations.

Another one that I have in mind is implementing kind of specific variant of end-2-end encryption on top of passthru mode.
We have a predefined e2ee scheme in spec (still not finished yet). But what if the client/developer wants to go its own way? And there is no way to use payload dict for meta information as it is encrypted :) But for example, there is a need to pass some params, idk, iterations count for the PBKDF scheme or keylen. Then custom attributes can help! It is not a payload itself but rather additional meta information for implementing some more high-level communications/protocols.

What do you think folks?

**Comment by KSDaemon** [comment]
> Was it already discussed whether the router should propagate these custom attributes when doing CALL -> INVOCATION and YIELD -> RESULT?

All I can remember — is the discussion in this thread. Anyway, this definitely should be stated in the spec.

**Comment by ecorm** [comment]
I think the router should propagate a custom attribute, unless it understands it and knows it's supposed to consume it without propagation.

**Comment by KSDaemon** [comment]
Yeah. Agree. So if a router doesn't understand custom attributes — it should not discard them but instead just forward them to other peers.

It is okay if only the caller and the callee understand and expect some custom attributes, the router is outside of this business.

**Comment by ecorm** [comment]
With the `custom|dict`, implementations could further subdivide it with the name of their implementation to avoid clashes. For example:

```
CALL.Options.custom.cppwamp.foo
```

And if another implementation is used for the router, it could also put its custom attributes in the implementation's "namespace":

```
INVOCATION.Details.custom = {"cppwamp.foo": 123, "crossbar.bar": 456}
```

The same could be done with `x_` prefixes:

```
INVOCATION.Details = {"x_cppwamp.foo": 123, "x_crossbar.bar": 456}
```

**Comment by KSDaemon** [comment]
@ecorm that were short versions of:

`x_*` + allow prefixed attributes from client side → End user must provide custom attributes already prefixed and client library only checks them for x_ prefix

`x_*` + allow unprefixed attributes from client side → End user provides unprefixed attribute and client library automatically adds x_ prefix

**Comment by KSDaemon** [comment]
I thought that these are really implementation details and probably should not be described in the specification

**Comment by ecorm** [comment]
I prefer option 4 (`custom|dict`), as it's easier for the router/receiver to check for a single `custom` key than to scan for all keys starting with `x_`.

If there are implementations out there that are already using `x_`, then I guess we should consider "grandfathering" `x_` into the spec as well.

> `x_*` + allow unprefixed attributes from client side → End user provides unprefixed attribute and client library automatically adds x_ prefix

This should be a library "convenience" feature and not part of the spec.

**Comment by oberstet** [comment]
just a quick note as "preface"

> I don't like magic happening behind my eyes.

yeah, agreed! like "principle of least surprise" and "batteries included", a principle of "no hidden magic" ..

now, regarding the options listed by @KSDaemon I'm wondering whether I just don't get it, or whether this is already one of the options:

* allow any `x_` prefixed top-level attributes in `Details` and `Options`, both from client and router side; any peer must forward `x_` without modification

?

**Comment by KSDaemon** [comment]
@oberstet Yeah, mb that is not stated clearly but it's kinda 1st option. Allow `x_*` top level attributes.

> Yeah. Agree. So if a router doesn't understand custom attributes — it should not discard them but instead just forward them to other peers.
It is okay if only the caller and the callee understand and expect some custom attributes, the router is outside of this business.

**Comment by aramallo** [comment]
I think supporting the prefix  `x_` and the `custom` dict is a nice option. 

The only issue is that we might need to define a conflict resolution rule when the user passes the same attribute under both mechanisms.

Finally we'll need to explain in spec that custom attributes override standard attributes as discussed before. I.e `Call.Options.invoke` is overridden by `Call.Options.x_invoke` or `Call.Options.custom.invoke`.

**Comment by oberstet** [comment]
my view rgd custom dict: unnecessary (if we have `x_` top level attributes)

> custom attributes override standard attributes

-1

**Comment by KSDaemon** [comment]
> The only issue is that we might need to define a conflict resolution rule when the user passes the same attribute under both mechanisms.

I don't think there is a conflict here. It's up to the client to do and resolve it in whatever way. Wamp peers do not process these attributes in any way, just pass aka proxy them to the client-side code, or in case the router understands it — it can process them in its own defined fashion. And authors of a such router may describe in their documentation how all related aspects (including conflict resolution) work.

In my view, this also means that there is no overriding that should be described in spec, it is implementation specific. 

So having the example by @aramallo with `invoke` / `x_invoke`: 

* All routers that do not understand `x_invoke` — will just pass this attribute within the `INVOCATION` message
* Routers that DO understand it — may process them as they want. But my opinion is that:
  * Any custom attributes SHOULD or even MUST not override the official one. Well with one exception: if the router implements a new invocation policy, that is not defined officially by the spec, then I think the client should set `x_invoke` to this specific value, and at the same time set `invoke` to one of the supported values (or omit and thus having `single` as the default value) as the fallback and clearly describe in the router documentation that `x_invoke` if specified, will have precedence over `invoke`. This preserves router interoperability. 

I mean I think it is important to preserve interoperability between clients and routers. And if peers implement some custom logic that modifies or extends the official feature — there should always be a fallback to official behavior. Only with this approach we can evolve and grow the WAMP ecosystem.

**Comment by ecorm** [comment]
> The only issue is that we might need to define a conflict resolution rule when the user passes the same attribute under both mechanisms.

I now feel more strongly that we should drop the `x_foo` mechanism due to the possible performance hit of having to scan for this prefix. Checking for a single `custom` sub-dictionary is way simpler and more efficient. Especially in a language supporting move semantics.

**Comment by KSDaemon** [comment]
That was my initial thoughts when proposing a sub-dictionary variant :)
